### ISSUE TITLE
Not able to scroll the page

### ISSUE THREAD

**Issue by rscwcd** [mozilla/geckodriver#776]
Hi, 

After upgrading our project to use gecko driver(Version "geckodriver-v0.16.0-win64") and FF 53.0, found that a feature to scroll down the page is not working. Tried many different techniques but didn't find any way to scroll the page to go through required element. We are using below code to go through this:

   public T scroll_to_element(By element) {
        WebElement webElement = findElement(element);
        Actions actions = new Actions(driver);
        actions.moveToElement(webElement).perform();
        return me();
    }

Please advise.

**Comment by andreastt** [comment]
The `actions.moveToElement` API is a Selenium API and does not correspond to a primitive in the WebDriver specification. I suspect there is a bug with the shim Selenium provide for this.

Could you open a bug with Selenium, and I will keep this open for tracking purposes?

**Comment by rscwcd** [comment]
@andreastt : Hi, I have generated a new issue on Selenium Forum with below link 

[actions.moveToElement in Selenium does not scroll down the page #4148](https://github.com/SeleniumHQ/selenium/issues/4148)

**Comment by mseele** [comment]
Is this really a bug in selenium? I run into that issue and the same code in chrome works for me.

Here is the output and stacktrace i got with geckodriver 0.18 and ff55:

```
1502259066277	webdriver::server	DEBUG	-> POST /session/8a024dc4-2006-4bf0-a90f-6a90f98cbafd/actions {"actions":[{"id":"default mouse","type":"pointer","parameters":{"pointerType":"mouse"},"actions":[{"duration":100,"x":0,"y":0,"type":"pointerMove","origin":{"ELEMENT":"4be391ab-d36f-4329-b13c-7a7eba146b3d","element-6066-11e4-a52e-4f735466cecf":"4be391ab-d36f-4329-b13c-7a7eba146b3d"}}]}]}
1502259066277	geckodriver::marionette	TRACE	-> 265:[0,446,"performActions",{"actions":[{"actions":[{"duration":100,"origin":{"element-6066-11e4-a52e-4f735466cecf":"4be391ab-d36f-4329-b13c-7a7eba146b3d"},"type":"pointerMove","x":0,"y":0}],"id":"default mouse","parameters":{"pointerType":"mouse"},"type":"pointer"}]}]
1502259066278	Marionette	TRACE	0 -> [0,446,"performActions",{"actions":[{"actions":[{"duration":100,"origin":{"element-6066-11e4-a52e-4f735466cecf":"4be391ab-d36f-4329-b13c-7a7eba146b3d"},"type":"pointerMove","x":0,"y":0}],"id":"default mouse","parameters":{"pointerType":"mouse"},"type":"pointer"}]}]
1502259066285	Marionette	TRACE	0 <- [1,446,{"error":"move target out of bounds","message":"(-346, 836) is out of bounds of viewport width (1920) and height (1091)","stacktrace":"WebDriverError@chrome://marionette/content/error.js:222:5\nMoveTargetOutOfBoundsError@chrome://marionette/content/error.js:416:5\ndispatchPointerMove/<@chrome://marionette/content/action.js:1247:13\ndispatchPointerMove@chrome://marionette/content/action.js:1240:10\ntoEvents/<@chrome://marionette/content/action.js:1065:16\naction.dispatchTickActions@chrome://marionette/content/action.js:974:23\naction.dispatch/chainEvents<@chrome://marionette/content/action.js:944:13\nTaskImpl_run@resource://gre/modules/Task.jsm:321:42\nTaskImpl@resource://gre/modules/Task.jsm:279:3\nasyncFunction@resource://gre/modules/Task.jsm:254:14\nTask_spawn@resource://gre/modules/Task.jsm:168:12\naction.dispatch@chrome://marionette/content/action.js:942:21\nperformActions@chrome://marionette/content/listener.js:849:9\nTaskImpl_run@resource://gre/modules/Task.jsm:321:42\nTaskImpl@resource://gre/modules/Task.jsm:279:3\nasyncFunction@resource://gre/modules/Task.jsm:254:14\nTask_spawn@resource://gre/modules/Task.jsm:168:12\nTaskImpl_handleResultValue@resource://gre/modules/Task.jsm:391:16\nTaskImpl_run@resource://gre/modules/Task.jsm:329:15\nTaskImpl@resource://gre/modules/Task.jsm:279:3\nasyncFunction@resource://gre/modules/Task.jsm:254:14\nTask_spawn@resource://gre/modules/Task.jsm:168:12\ndispatch/<@chrome://marionette/content/listener.js:427:15\n"},null]
1502259066277	geckodriver::marionette	TRACE	<- [1,446,{"error":"move target out of bounds","message":"(-346, 836) is out of bounds of viewport width (1920) and height (1091)","stacktrace":"WebDriverError@chrome://marionette/content/error.js:222:5\nMoveTargetOutOfBoundsError@chrome://marionette/content/error.js:416:5\ndispatchPointerMove/<@chrome://marionette/content/action.js:1247:13\ndispatchPointerMove@chrome://marionette/content/action.js:1240:10\ntoEvents/<@chrome://marionette/content/action.js:1065:16\naction.dispatchTickActions@chrome://marionette/content/action.js:974:23\naction.dispatch/chainEvents<@chrome://marionette/content/action.js:944:13\nTaskImpl_run@resource://gre/modules/Task.jsm:321:42\nTaskImpl@resource://gre/modules/Task.jsm:279:3\nasyncFunction@resource://gre/modules/Task.jsm:254:14\nTask_spawn@resource://gre/modules/Task.jsm:168:12\naction.dispatch@chrome://marionette/content/action.js:942:21\nperformActions@chrome://marionette/content/listener.js:849:9\nTaskImpl_run@resource://gre/modules/Task.jsm:321:42\nTaskImpl@resource://gre/modules/Task.jsm:279:3\nasyncFunction@resource://gre/modules/Task.jsm:254:14\nTask_spawn@resource://gre/modules/Task.jsm:168:12\nTaskImpl_handleResultValue@resource://gre/modules/Task.jsm:391:16\nTaskImpl_run@resource://gre/modules/Task.jsm:329:15\nTaskImpl@resource://gre/modules/Task.jsm:279:3\nasyncFunction@resource://gre/modules/Task.jsm:254:14\nTask_spawn@resource://gre/modules/Task.jsm:168:12\ndispatch/<@chrome://marionette/content/listener.js:427:15\n"},null]
1502259066277	webdriver::server	DEBUG	<- 500 Internal Server Error {"value":{"error":"move target out of bounds","message":"(-346, 836) is out of bounds of viewport width (1920) and height (1091)","stacktrace":"stack backtrace:\n   0:           0x4bb74f - <no info>\n   1:           0x4bbea9 - <no info>\n   2:           0x43ce8d - <no info>\n   3:           0x44ce14 - <no info>\n   4:           0x44944a - <no info>\n   5:           0x4203e1 - <no info>\n   6:           0x407dc7 - <no info>\n   7:           0x6d95b9 - <no info>\n   8:           0x4173a7 - <no info>\n   9:           0x6d38b3 - <no info>\n  10:     0x7ff9fbce2774 - BaseThreadInitThunk"}}


org.openqa.selenium.interactions.MoveTargetOutOfBoundsException: (-346, 836) is out of bounds of viewport width (1920) and height (1091)
Build info: version: '3.4.0', revision: 'unknown', time: 'unknown'
System info: host: 'GHCL11', ip: '192.168.100.116', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_144'
Driver info: org.openqa.selenium.firefox.FirefoxDriver
Capabilities [{moz:profile=C:\Users\mseele\AppData\Local\Temp\rust_mozprofile.zjj4p54AWmwC, rotatable=false, timeouts={implicit=0.0, pageLoad=300000.0, script=30000.0}, pageLoadStrategy=normal, platform=ANY, specificationLevel=0.0, moz:accessibilityChecks=false, acceptInsecureCerts=true, browserVersion=55.0, platformVersion=10.0, moz:processID=12552.0, browserName=firefox, javascriptEnabled=true, platformName=windows_nt}]
Session ID: 8a024dc4-2006-4bf0-a90f-6a90f98cbafd
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.openqa.selenium.remote.http.W3CHttpResponseCodec.createException(W3CHttpResponseCodec.java:150)
	at org.openqa.selenium.remote.http.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:115)
	at org.openqa.selenium.remote.http.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:45)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:164)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:82)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:637)
	at org.openqa.selenium.remote.RemoteWebDriver.perform(RemoteWebDriver.java:703)
	at org.openqa.selenium.interactions.Actions$BuiltAction.perform(Actions.java:631)
	at org.openqa.selenium.interactions.Actions.perform(Actions.java:593)
```

This is the code i execute:
```	
/**
 * Click at the given web element position. Please use that method instead
 * of the {@link Actions#click()} method.
 * */
public void mouseClick(WebDriver driver, WebElement element) {
	if (element != null) {
		System.err.println("MOVE TO ELEMENT > ");
		new Actions(driver).moveToElement(element).perform();
		System.err.println("< MOVE TO ELEMENT");
	}
	System.err.println("CLICK");
	// we need to do a clickAndHold and a release cause the normal click
	// does not work under firefox
	// (https://code.google.com/p/selenium/issues/detail?id=3578)
	new Actions(driver).clickAndHold().release().perform();
}
```

**Comment by gerhardj** [comment]
@mseele I would say it depends on the interpretation of the specs. Chrome might do an implicit move on a mouse click, while Firefox doesn't. But if you strictly follow the specs, there is no scroll demanded. Therefore I would consider this an upstream bug/missing feature. It has been confirmed by upstream in the linked issue:

w3c/webdriver#1005

**Comment by mseele** [comment]
I don't see this as "solution". My problem ist: all the "other" browsers (chrome, ie, edge) are working with this code. Firefox worked with the "old" driver and with geckodriver it does not work anymore.

There is no "code" to scroll an element which is out of the viewport into the viewport with geckodriver. Because of this, imho this is a bug for me. How to solve the problem of scrolling an element which is out of the viewport into the viewport with geckodriver?

**Comment by SankySyS** [comment]
Workaround on Java:

```
WebElement element = driver.findElement(locator);
((JavascriptExecutor) driver).executeScript("arguments[0].scrollIntoView(true);", element);
```

**Comment by whimboo** [comment]
Once the spec has been updated, Marionette will get that fix via [bug 1295538](https://bugzilla.mozilla.org/show_bug.cgi?id=1295538)

**Comment by mseele** [comment]
Any progress going on here? [bug 1295538](https://bugzilla.mozilla.org/show_bug.cgi?id=1295538) has been closed as wontfix. 
I cannot believe that i need this workaround
```
WebElement element = driver.findElement(locator);
((JavascriptExecutor) driver).executeScript("arguments[0].scrollIntoView(true);", element);
```
for firefox and all the other browsers just work out of the box.

**Comment by whimboo** [comment]
Well, as said this basically needs the spec updated first (https://github.com/w3c/webdriver/issues/1005). Once this is done we might reopen bug 1295538 to get this fixed.
